parquet_file/src/catalog.rs

```diff
@@ -0,0 +1,1237 @@
+        let path = transaction_path(object_store, server_id, db_name, tkey);
+        let proto = load_transaction_proto(object_store, &path).await?;
+
+        // sanity-check file content
+        if proto.version != TRANSACTION_VERSION {
+            TransactionVersionMismatch {
+                revision_counter: tkey.revision_counter,
+                actual: proto.version,
+                // we only support a single version right now
+                expected: vec![TRANSACTION_VERSION],
+            }
+            .fail()?;
+        }
+        if proto.revision_counter != tkey.revision_counter {
+            WrongTransactionRevision {
+                actual: proto.revision_counter,
+                expected: tkey.revision_counter,
+            }
+            .fail()?
+        }
+        let uuid_actual = parse_uuid_required(&proto.uuid)?;
+        if uuid_actual != tkey.uuid {
+            WrongTransactionUuid {
+                revision_counter: tkey.revision_counter,
+                expected: tkey.uuid,
+                actual: uuid_actual,
+            }
+            .fail()?
+        }
+        let last_uuid_actual = parse_uuid(&proto.previous_uuid)?;
+        let last_uuid_expected = last_tkey.as_ref().map(|tkey| tkey.uuid);
+        if last_uuid_actual != last_uuid_expected {
+            WrongTransactionLink {
+                revision_counter: tkey.revision_counter,
+                expected: last_uuid_expected,
+                actual: last_uuid_actual,
+            }
+            .fail()?;
+        }
+
+        // apply
+        for action in &proto.actions {
+            if let Some(action) = action.action.as_ref() {
+                Self::handle_action(&mut state, action)?;
+            }
+        }
+
+        Ok(Self {
+            proto,
+            next_state: state,
+        })
+    }
+}
+
+/// Handle for an open uncommitted transaction.
+///
+/// Dropping this object w/o calling [`commit`](Self::commit) will issue a warning.
+pub struct TransactionHandle<'c> {
+    catalog: &'c mut PreservedCatalog,
+    transaction: Option<OpenTransaction>,
+}
+
+impl<'c> TransactionHandle<'c> {
+    fn new(catalog: &'c mut PreservedCatalog) -> Self {
+        let transaction = OpenTransaction::new(catalog);
+        let tkey = transaction.tkey();
+        info!(?tkey, "transaction started");
+
+        Self {
+            catalog,
+            transaction: Some(transaction),
+        }
+    }
+
+    /// Write data to object store and commit transaction to underlying catalog.
+    pub async fn commit(mut self) -> Result<()> {
+        // write to object store
+        self.transaction
+            .as_mut()
+            .unwrap()
+            .store(
+                &self.catalog.object_store,
+                self.catalog.server_id,
+                &self.catalog.db_name,
+            )
+            .await?;
+
+        // commit to catalog
+        let mut t = None;
+        std::mem::swap(&mut t, &mut self.transaction);
```

Comment:
`mem::take()` is great for this!

```suggestion
        let t = std::mem::take(&mut self.transaction);
```